File: shared/agent-core/src/goals.ts
import { Database } from "bun:sqlite";
import type { Goal, ProjectCode } from "./types";

const CREATE_TABLE = `CREATE TABLE IF NOT EXISTS agent_goals (
  id TEXT PRIMARY KEY,
  title TEXT NOT NULL,
  description TEXT NOT NULL,
  projects TEXT NOT NULL,
  status TEXT DEFAULT 'active' CHECK(status IN ('active','completed','paused')),
  priority TEXT DEFAULT 'medium' CHECK(priority IN ('high','medium','low')),
  deadline TEXT,
  progress TEXT,
  created_at TEXT NOT NULL,
  updated_at TEXT NOT NULL
)`;

export class GoalStore {
  private db: Database;
  constructor(db: Database) {
    this.db = db;
    this.db.run(CREATE_TABLE);
  }

  createGoal(params: { title: string; description: string; projects: ProjectCode[]; priority?: "high"|"medium"|"low"; deadline?: string }): Goal {
    const id = crypto.randomUUID();
    const now = new Date().toISOString();
    const goal: Goal = {
      id, title: params.title, description: params.description,
      projects: params.projects, status: "active",
      priority: params.priority || "medium",
      deadline: params.deadline || null, progress: null,
      created_at: now, updated_at: now,
    };
    this.db.run(
      `INSERT INTO agent_goals (id,title,description,projects,status,priority,deadline,progress,created_at,updated_at) VALUES (?,?,?,?,?,?,?,?,?,?)`,
      [goal.id, goal.title, goal.description, JSON.stringify(goal.projects), goal.status, goal.priority, goal.deadline, goal.progress, goal.created_at, goal.updated_at]
    );
    return goal;
  }

  getAll(): Goal[] {
    return (this.db.query("SELECT * FROM agent_goals ORDER BY created_at DESC").all() as any[]).map(rowToGoal);
  }

  getActive(): Goal[] {
    return (this.db.query("SELECT * FROM agent_goals WHERE status = 'active' ORDER BY priority ASC, created_at DESC").all() as any[]).map(rowToGoal);
  }

  getByProject(project: ProjectCode): Goal[] {
    // projects is a JSON array, use LIKE to find goals that include this project
    return (this.db.query("SELECT * FROM agent_goals WHERE status = 'active' AND projects LIKE ? ORDER BY priority ASC").all(`%"${project}"%`) as any[]).map(rowToGoal);
  }

  getById(id: string): Goal | null {
    const row = this.db.query("SELECT * FROM agent_goals WHERE id = ?").get(id) as any;
    return row ? rowToGoal(row) : null;
  }

  updateGoal(id: string, updates: Partial<Pick<Goal, "title"|"description"|"projects"|"priority"|"deadline"|"progress"|"status">>): Goal | null {
    const goal = this.getById(id);
    if (!goal) return null;
    const now = new Date().toISOString();
    if (updates.title !== undefined) goal.title = updates.title;
    if (updates.description !== undefined) goal.description = updates.description;
    if (updates.projects !== undefined) goal.projects = updates.projects;
    if (updates.priority !== undefined) goal.priority = updates.priority;
    if (updates.deadline !== undefined) goal.deadline = updates.deadline;
    if (updates.progress !== undefined) goal.progress = updates.progress;
    if (updates.status !== undefined) goal.status = updates.status;
    goal.updated_at = now;
    this.db.run(
      `UPDATE agent_goals SET title=?,description=?,projects=?,priority=?,deadline=?,progress=?,status=?,updated_at=? WHERE id=?`,
      [goal.title, goal.description, JSON.stringify(goal.projects), goal.priority, goal.deadline, goal.progress, goal.status, goal.updated_at, id]
    );
    return goal;
  }

  deleteGoal(id: string): boolean {
    const result = this.db.run("DELETE FROM agent_goals WHERE id = ?", [id]);
    return result.changes > 0;
  }
}

function rowToGoal(row: any): Goal {
  return {
    ...row,
    projects: JSON.parse(row.projects),
  };
}

File: content-pipeline/backend/src/agent/goals.py
"""GoalStore — Python port of shared/agent-core/src/goals.ts"""

import json
import uuid
from datetime import datetime

from src.db.connection import Database

PROJECT_CODES = {"COIN", "BLOG", "DASH", "TRAIN", "TGBOT", "DCBOT", "GENERAL"}


class GoalStore:
    def __init__(self, db: Database):
        self.db = db

    def create_goal(
        self,
        title: str,
        description: str,
        projects: list[str],
        priority: str = "medium",
        deadline: str | None = None,
    ) -> dict:
        goal_id = str(uuid.uuid4())
        now = datetime.now().isoformat()
        with self.db._lock:
            self.db.conn.execute(
                "INSERT INTO agent_goals (id,title,description,projects,status,priority,deadline,progress,created_at,updated_at) "
                "VALUES (?,?,?,?,?,?,?,?,?,?)",
                (goal_id, title, description, json.dumps(projects), "active", priority, deadline, None, now, now),
            )
            self.db.conn.commit()
        return self.get_by_id(goal_id)  # type: ignore

    def get_all(self) -> list[dict]:
        with self.db._lock:
            rows = self.db.conn.execute(
                "SELECT * FROM agent_goals ORDER BY created_at DESC"
            ).fetchall()
        return [_row_to_goal(r) for r in rows]

    def get_active(self) -> list[dict]:
        with self.db._lock:
            rows = self.db.conn.execute(
                "SELECT * FROM agent_goals WHERE status = 'active' ORDER BY priority ASC, created_at DESC"
            ).fetchall()
        return [_row_to_goal(r) for r in rows]

    def get_by_project(self, project: str) -> list[dict]:
        with self.db._lock:
            rows = self.db.conn.execute(
                "SELECT * FROM agent_goals WHERE status = 'active' AND projects LIKE ? ORDER BY priority ASC",
                (f'%"{project}"%',),
            ).fetchall()
        return [_row_to_goal(r) for r in rows]

    def get_by_id(self, goal_id: str) -> dict | None:
        with self.db._lock:
            row = self.db.conn.execute(
                "SELECT * FROM agent_goals WHERE id = ?", (goal_id,)
            ).fetchone()
        return _row_to_goal(row) if row else None

    def update_goal(self, goal_id: str, updates: dict) -> dict | None:
        goal = self.get_by_id(goal_id)
        if not goal:
            return None
        now = datetime.now().isoformat()
        fields = []
        params = []
        for key in ("title", "description", "priority", "deadline", "progress", "status"):
            if key in updates:
                fields.append(f"{key} = ?")
                params.append(updates[key])
        if "projects" in updates:
            fields.append("projects = ?")
            params.append(json.dumps(updates["projects"]))
        if not fields:
            return goal
        fields.append("updated_at = ?")
        params.append(now)
        params.append(goal_id)
        with self.db._lock:
            self.db.conn.execute(
                f"UPDATE agent_goals SET {', '.join(fields)} WHERE id = ?", params
            )
            self.db.conn.commit()
        return self.get_by_id(goal_id)

    def delete_goal(self, goal_id: str) -> bool:
        with self.db._lock:
            cursor = self.db.conn.execute("DELETE FROM agent_goals WHERE id = ?", (goal_id,))
            self.db.conn.commit()
        return cursor.rowcount > 0


def _row_to_goal(row) -> dict:
    d = dict(row)
    d["projects"] = json.loads(d["projects"])
    return d
